### PR TITLE
Filter instrumentation and bytecode upgrade interceptors

### PR DESCRIPTION
**Pull request by asodja**
Now interceptors can be filtered with:
- `BytecodeInterceptorRequest.INSTRUMENTATION_ONLY` -> instrumentation only
- `BytecodeInterceptorRequest.ALL` -> instrumentation + bytecode upgrades

There are two main parts:

1. Annotation processor changes: Now you can add type of interceptor to the class, e.g.:
```
@SpecificJvmCallInterceptors(generatedClassName = "SomeName", type = BytecodeInterceptorType.INSTRUMENTATION)
class MyInterceptorDeclarations {}
```
By default interceptor is of type `INSTRUMENTATION`. Annotation processor then adds a marker interface to the interceptor, so we can filter interceptors at Gradle instrumentation. 

Additionally you can now also use `@InjectVisitorContext` and inject `BytecodeInterceptorRequest` as parameter. That is needed only to instrument `ScriptBytecodeAdapter`, so there is no other use case.
See changes in `:internal-instrumentation-processor`.

2. Gradle runtime changes:
We can now filter interceptors, by passing `BytecodeInterceptorRequest` to `InstrumentingClassTransform`. See changes in `:core`.

This request is then passed down to different handle methods so we can filter interceptors. Ideally we would have some kind of a "context" object here, but since we have to generate bytecode with such parameter, it's a bit easier to just use enum for now. But feel free to suggest a different approach. 

Fixes https://github.com/gradle/gradle/issues/26396

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java [hunk 1]
```diff
@@ -361,10 +385,25 @@ private static void validateSignature(CallableInfo callable) {
...
+        boolean hasInjectVisitorContext = hasInjectVisitorContext(callable);
+        if (hasInjectVisitorContext) {
+            ParameterInfo lastParameter = callable.getParameters().get(callable.getParameters().size() - 1);
+            if (lastParameter.getKind() != ParameterKindInfo.INJECT_VISITOR_CONTEXT) {
+                throw new Failure("The interceptor's @" + ParameterKind.InjectVisitorContext.class.getSimpleName() + " parameter should be last parameter");
+            }
+            if (!lastParameter.getParameterType().getClassName().equals(BytecodeInterceptorRequest.class.getName())) {
+                throw new Failure("The interceptor's @" + ParameterKind.InjectVisitorContext.class.getSimpleName() + " parameter should be of type " + BytecodeInterceptorRequest.class.getName() + " but was " + lastParameter.getParameterType().getClassName());
+            }
+            if (callable.getParameters().stream().filter(it -> it.getKind() == ParameterKindInfo.INJECT_VISITOR_CONTEXT).count() > 1) {
```
Comment (h0tk3y) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java:397]: 👍

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/types/BytecodeInterceptorRequest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.EnumSet;
+import java.util.Set;
+
+/**
+ * Request for interceptors. Currently, we support filter just for interception types, but in the feature we could also filter per version etc.
+ *
+ * Implemented as a enum, so it's easier to generate bytecode.
+ */
+public enum BytecodeInterceptorRequest {
```
Comment (h0tk3y) [inline platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/types/BytecodeInterceptorRequest.java:27]: 🤔 Would it make sense to find a name that would better reflect that this entity does not contain any real "payload" of the request? Looking at the word "request" at the use sites made me think about more complex data structure being involved. Maybe something about interception context filter?

**Review by h0tk3y — APPROVED** [review]
LGTM 👍
I wonder if there is anything that we can do to simplify adding new kinds of such "extras" that interceptors use at runtime...

**Review by h0tk3y — APPROVED** [review]
LGTM 👍
I wonder if there is anything that we can do to simplify adding new kinds of such "extras" that interceptors use at runtime...

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/types/BytecodeInterceptorRequest.java (hunk 2 above)
Comment (asodja) [reply platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/types/BytecodeInterceptorRequest.java:27]: Yes, I can rename that. Maybe we can go with BytecodeInterceptorFilter. And if we will add more context later maybe we can rename it

**Comment by asodja** [comment]
> I wonder if there is anything that we can do to simplify adding new kinds of such "extras" that interceptors use at runtime...

Good point, I agree we could simplify it. I think we could do that when we have another usecase to inject something in to the interceptors.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle merge

**Pull request merged**